JIT on Windows

Opening this issue to track the problems preventing JIT on Windows from happening yet.

From my own experience, the problem is due to the difference between Windows calling convention and Sys-V.

- [x86_64 Windows](https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170) and x86_64 Sys-V passes arguments in different registers. A solution might be just keep using the Sys-V ABI in the JITted code and make a trampoline before returning to compiled Windows code.
- FS - GS register differences. Although as JIT already works on macOS (which also uses different FS/GS than Linux), I guess the same can be applied to Windows?
- x86_64 Sys-V requires a 128-byte red zone below `rsp`, while x86_64 Windows does not. When jumping to JIT code, the value of `rsp` must be 128 bytes lower than it is supposed to be, and all read/write operations involving `rsp` must be hooked to modify the value. Things will get tricky when dealing with opcodes such as `call`,...

Don't know if this is related to the JIT, or more related to #26, but tests are still failing:

```
o//blink/blink third_party/cosmo/2/sigaction_test.com
I2023-01-17T03:05:07.743414:blink/syscall.c:2261:3638 unrecognized sigaction() flags: 0x2
error:test/libc/calls/sigaction_test.c:200: sigaction_autoZombieSlayer() on fv-az449-110 pid 3638 tid 3638
	ASSERT_EQ(0, sigaction(SIGCHLD, &sa, &sa))
		need 0	(or 0 or  =
		 got -1	(or 0xffffffffffffffff)
	EINVAL/22/Invalid argument
	third_party/cosmo/2/sigaction_test.com @ fv-az449-110
1 / 9 tests failed
make: *** [third_party/cosmo/cosmo.mk:20: o//third_party/cosmo/2/sigaction_test.com.ok] Error 1
make: *** Waiting for unfinished jobs....
```

The branch is based on unmodified code from `master`, no special configuration.

See the full logs [here](https://github.com/trungnt2910/blink/actions/runs/3935639818/jobs/6731502748).

--- a/blink/jit.c
+++ b/blink/jit.c
@@ -386,6 +386,22 @@ int DisableJit(struct Jit *jit) {
   return 0;
 }
 
+/**
+ * Fixes the memory protection for existing Just-In-Time code blocks.
+ */
+int FixJitProtection(struct Jit *jit) {
+  struct Dll *e;
+  LOCK(&jit->lock);
+  int prot;
+  prot = atomic_load_explicit(&g_jit.prot, memory_order_relaxed);
+  for (e = dll_first(jit->blocks); e; e = dll_next(jit->blocks, e)) {
+    unassert(!Mprotect(JITBLOCK_CONTAINER(e)->addr, jit->blocksize, prot,
+                       "jit"));
+  }
+  UNLOCK(&jit->lock);
+  return 0;
+}
+
 bool SetJitHook(struct Jit *jit, u64 virt, intptr_t func) {
   int offset;
   unsigned i;
@@ -1113,6 +1129,7 @@ STUB(int, InitJit, (struct Jit *jit), 0)
 STUB(int, DestroyJit, (struct Jit *jit), 0)
 STUB(int, ShutdownJit, (void), 0)
 STUB(int, DisableJit, (struct Jit *jit), 0)
+STUB(int, FixJitProtection, (struct Jit *jit), 0)
 STUB(bool, AppendJit, (struct JitBlock *jb, const void *data, long size), 0)
 STUB(bool, AppendJitMovReg, (struct JitBlock *jb, int dst, int src), 0)
 STUB(int, AbandonJit, (struct Jit *jit, struct JitBlock *jb), 0)

--- a/blink/jit.h
+++ b/blink/jit.h
@@ -155,6 +155,7 @@ int ShutdownJit(void);
 int InitJit(struct Jit *);
 int DestroyJit(struct Jit *);
 int DisableJit(struct Jit *);
+int FixJitProtection(struct Jit *);
 bool CanJitForImmediateEffect(void) nosideeffect;
 bool AppendJit(struct JitBlock *, const void *, long);
 int AbandonJit(struct Jit *, struct JitBlock *);

--- a/blink/syscall.c
+++ b/blink/syscall.c
@@ -368,10 +368,9 @@ static int SysFork(struct Machine *m) {
     m->system->isfork = true;
     RemoveOtherThreads(m->system);
 #ifdef __CYGWIN__
-    LOGF("disabling jit due to fork");
-    DestroyJit(&m->system->jit);
-    InitJit(&m->system->jit);
-    DisableJit(&m->system->jit);
+    // Cygwin doesn't seem to properly set the PROT_EXEC
+    // protection for JIT blocks after forking.
+    FixJitProtection(&m->system->jit);
 #endif
   }
   return pid;

